packages/langium/src/utils/grammar-util.ts

```diff
@@ -54,26 +54,38 @@ export function getAllReachableRules(grammar: ast.Grammar, allTerminals: boolean
     }
 
     const topMostRules = [entryRule as ast.AbstractRule].concat(getHiddenRules(grammar));
+    const rules = new Set<ast.AbstractRule>();
     for (const rule of topMostRules) {
-        ruleDfs(rule, ruleNames, allTerminals);
+        ruleDfs(rule, ruleNames, rules, allTerminals);
     }
 
-    const rules = new Set<ast.AbstractRule>();
+    
     for (const rule of grammar.rules) {
-        if (ruleNames.has(rule.name) || (ast.isTerminalRule(rule) && rule.hidden)) {
+        if ((ast.isTerminalRule(rule) && rule.hidden)) {
             rules.add(rule);
         }
     }
     return rules;
 }
 
-function ruleDfs(rule: ast.AbstractRule, visitedSet: Set<string>, allTerminals: boolean): void {
-    visitedSet.add(rule.name);
+function ruleDfs(rule: ast.AbstractRule, visitedRuleNames: Set<string>, visitedRules: Set<ast.AbstractRule> , allTerminals: boolean): void {
+    visitedRuleNames.add(rule.name);
+    visitedRules.add(rule)
     streamAllContents(rule).forEach(node => {
         if (ast.isRuleCall(node) || (allTerminals && ast.isTerminalRuleCall(node))) {
             const refRule = node.rule.ref;
-            if (refRule && !visitedSet.has(refRule.name)) {
-                ruleDfs(refRule, visitedSet, allTerminals);
+            if (refRule && !visitedRuleNames.has(refRule.name)) {
+                ruleDfs(refRule, visitedRuleNames, visitedRules, allTerminals);
+            }
+        } else if (ast.isCrossReference(node)) {
+            const term = getCrossReferenceTerminal(node)
```

Comment:
what is the override semantic here: rule used at the name place is overriden/also defined in current grmmar

Reply:
Langium prevents adopters from overriding imported rules.

Reply:
ok, need to try that out